From 813377ba95d28a4db493da54354fda82d89af9c1 Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Thu, 18 Jan 2018 18:00:24 -0800
Subject: [PATCH 09/10] dmaengine: xilinx: dpdma: Rechedule errored descriptor

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

If there's an error, reschedule the descriptor by adding it
back to submitted descriptor. If there's already pending or
submitted descriptor, drop and free the error descriptor properly.
Previously, the dropped descriptor wasn't freed properly.

Reported-by: Varunkumar Allagadapa <varunkum@xilinx.com>
Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Tested-by: Varunkumar Allagadapa <varunkum@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
(cherry picked from commit 83fde7040e20eb00eefcdd5f41e8f73d4334ca72)
---
 drivers/dma/xilinx/xilinx_dpdma.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/dma/xilinx/xilinx_dpdma.c b/drivers/dma/xilinx/xilinx_dpdma.c
index 146c0d2..d64c722 100644
--- a/drivers/dma/xilinx/xilinx_dpdma.c
+++ b/drivers/dma/xilinx/xilinx_dpdma.c
@@ -1742,9 +1742,12 @@ static void xilinx_dpdma_chan_handle_err(struct xilinx_dpdma_chan *chan)
 	case ACTIVE:
 	case PREPARED:
 		/* Reschedule if there's no new descriptor */
-		if (!chan->submitted_desc)
+		if (!chan->pending_desc && !chan->submitted_desc) {
+			chan->active_desc->status = ERRORED;
 			chan->submitted_desc = chan->active_desc;
-		chan->active_desc->status = ERRORED;
+		} else {
+			xilinx_dpdma_chan_free_tx_desc(chan, chan->active_desc);
+		}
 		break;
 	}
 	chan->active_desc = NULL;
-- 
1.8.5.2.233.g932f7e4

